README
RCDevs OpenOTP Plugin for RoundCube version 1.2.3
Copyright (c) 2010-2015 RCDevs SA, All rights reserved.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


CHANGELOG
1.2.3
	- Added Contextual authentication 
	  Change the LoginMode to LDAP-only for incoming requests 
	  from trusted devices on known IPs.
	- Added support to OpenOTP Software Token with Push Notif Authentication
	  Extend php soapclient to add timeout capabilities  
	- WebService API is now versioned
	- U2F javascript scripts updated
	- OTP Challenge input doesn't show anymore characters in clear text (type=password) 
1.2.2
	 - Fixed SSL issue with PHP 5.6+
	 - Enhanced FIDO U2F Library calls
1.2.1
 	 - Added support for OpenOTP v1.2.1
1.2.0
 	 - Added support for OpenOTP v1.2 and FIDO U2F authentication. 
1.1.0 
	 - Plugin has been completely re-writen.
	 - Changed the plugin behavior with OpenOTP Challenge-Response.
	   > The plugin displays an overlay on Challenge-Response session.
	   > No modification of the RoundCube login form.

1.0.2
     - Added support for OpenOTP v1.1 and SimpleLogin method.
       The plugin will now transparently support any OpenOTP Login Mode including:
       > LDAP only
       > OTP only
       > LDAP+OTP with user challenge-response

1.0.1
     - Added compatibility with OpenOTP 1.0.17.
     - Added support for location-based policies in WebADM 1.2.3 & OpenOTP 1.0.17.

1.0.0
     Initial public release.


openotp_authentication/openotp.php
<?php
/*
 RCDevs OpenOTP Plugin for RoundCube Webmail v1.2.3
 Copyright (c) 2010-2016 RCDevs, All rights reserved.
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
  
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/
class openotp {

	private $plugin;
	private $home;
	private $openotp_auth;
	private $server_url;
	private $client_id;
	private $default_domain;
	private $user_settings;                                                                           
	private $proxy_host;                                                                              
	private $proxy_port;                                                                              
	private $proxy_username;
	private $proxy_password;
	private $soap_client;
	private $context_name = 'OpenOTPContext';
	private $context_size = 32;
	private $context_time = 2500000;	

	public function __construct($openotp_plugin, $home=''){
		
        $this->plugin = $openotp_plugin->rc;
	    $this->home = $home;

		// load config		
		$this->openotp_auth = $this->plugin->config->get('openotp_auth');
		$this->server_url = $this->plugin->config->get('openotp_server_url');
		$this->client_id = $this->plugin->config->get('openotp_client_id');
		$this->default_domain = $this->plugin->config->get('openotp_default_domain');
		$this->user_settings = $this->plugin->config->get('openotp_user_settings');                                                                                   
		$this->proxy_host = $this->plugin->config->get('openotp_proxy_host');                                                                               
		$this->proxy_port = $this->plugin->config->get('openotp_proxy_port');                                                                               
		$this->proxy_username = $this->plugin->config->get('openotp_proxy_username');
		$this->proxy_password = $this->plugin->config->get('openotp_proxy_password');
	}
	
	public function checkFile($file)
	{
		if (!file_exists($this->home . '/'.$file)) {
			return false;
		}
		return true;
	}
	
	public function checkSOAPext()
	{
		if (!extension_loaded('soap')) {
			return false;
		}
		return true;
	}
	
	public function enableOpenotp_auth()
	{
		return $this->openotp_auth;
	}
	
	public function getContext_name()
	{
		return $this->context_name;
	}
	public function getContext_size()
	{
		return $this->context_size;
	}	
	public function getContext_time()
	{
		return $this->context_time;
	}	
		
	public function getServer_url()
	{
		return $this->server_url;
	}
		
	public function getDomain($username)
	{
		$pos = strpos($username, "\\");
		if ($pos) {
			$ret['domain'] = substr($username, 0, $pos);
			$ret['username'] = substr($username, $pos+1);
		} else {                                                                                                                      
			$ret = $this->default_domain;
		}
		return $ret;
	}
	
	public function getOverlay($otpChallenge, $u2fChallenge, $message, $username, $session, $timeout, $ldappw, $domain){
		$overlay = <<<EOT
		function addOpenOTPDivs(){
			var overlay_bg = document.createElement("div");
			overlay_bg.id = 'openotp_overlay_bg';
			overlay_bg.style.position = 'fixed'; 
			overlay_bg.style.top = '0'; 
			overlay_bg.style.left = '0'; 
			overlay_bg.style.width = '100%'; 
			overlay_bg.style.height = '100%'; 
			overlay_bg.style.background = 'grey';
			overlay_bg.style.zIndex = "9998"; 
			overlay_bg.style["filter"] = "0.9";
			overlay_bg.style["-moz-opacity"] = "0.9";
			overlay_bg.style["-khtml-opacity"] = "0.9";
			overlay_bg.style["opacity"] = "0.9";
		
			var tokenform = document.getElementsByName("_token")[0].value;
			var overlay = document.createElement("div");
			overlay.id = 'openotp_overlay';
			overlay.style.position = 'absolute'; 
			overlay.style.top = '165px'; 
			overlay.style.left = '50%'; 
			overlay.style.width = '280px'; 
			overlay.style.marginLeft = '-180px';
			overlay.style.padding = '65px 40px 50px 40px';
			overlay.style.background = 'url($this->home/openotp_banner.png) no-repeat top left #E4E4E4';
			overlay.style.border = '5px solid #545454';
			overlay.style.borderRadius = '10px';
			overlay.style.MozBorderRadius = '10px';
			overlay.style.WebkitBorderRadius = '10px';
			overlay.style.boxShadow = '1px 1px 12px #555555';
			overlay.style.WebkitBoxShadow = '1px 1px 12px #555555';
			overlay.style.MozBoxShadow = '1px 1px 12px #555555';
			overlay.style.zIndex = "9999"; 
			overlay.innerHTML = '<a style="position:absolute; top:-12px; right:-12px;" href="./" title="close"><img src="$this->home/openotp_closebtn.png"/></a>'
			+ '<style>'
			+ 'blink { -webkit-animation: blink 1s steps(5, start) infinite; -moz-animation:    blink 1s steps(5, start) infinite; -o-animation:      blink 1s steps(5, start) infinite; animation: blink 1s steps(5, start) infinite; }'
			+ '	@-webkit-keyframes blink { to { visibility: hidden; } }'
			+ '@-moz-keyframes blink { to { visibility: hidden; } }'
			+ '@-o-keyframes blink { to { visibility: hidden; } }'
			+ '@keyframes blink { to { visibility: hidden; } }'
			+ '</style>'				
			+ '<div style="background-color:red; margin:0 -40px 0; height:4px; width:360px; padding:0;" id="count_red"><div style="background-color:orange; margin:0; height:4px; width:360px; padding:0;" id="div_orange"></div></div>'
			+ '<form style="margin-top:30px;" action="./" name="login" id="login-form-otp"  method="POST">'
			+ '<input type="hidden" name="_token" value="'+tokenform+'">'
			+ '<input type="hidden" name="_task" value="login">'
			+ '<input type="hidden" name="_action" value="login">'
			+ '<input type="hidden" name="openotp_state" value="$session">'
			+ '<input type="hidden" name="openotp_domain" value="$domain">'
			+ '<input type="hidden" name="openotp_username" value="$username">'
			+ '<input type="hidden" name="openotp_ldappw" value="$ldappw">'
			+ '<table width="100%">'
			+ '<tr><td style="text-align:center; font-weight:bold; font-size:14px;">$message</td></tr>'
			+ '<tr><td id="timout_cell" style="text-align:center; padding-top:4px; font-weight:bold; font-style:italic; font-size:11px;">Timeout: <span id="timeout">$timeout seconds</span></td></tr>'
EOT;
	
			if( $otpChallenge || ( !$otpChallenge && !$u2fChallenge ) ){
			$overlay .= <<<EOT
			+ '<tr><td id="inputs_cell" style="text-align:center; padding-top:25px;"><input style="border:1px solid grey; background-color:white;" type="password" size=15 name="openotp_password">&nbsp;'
			+ '<input style="padding:3px 10px;" type="submit" value="Ok" class="button mainaction"></td></tr>'
EOT;
			}
			
			if( $u2fChallenge ){		
			$overlay .= "+ '<tr style=\"border:none;\"><td id=\"inputs_cell\" style=\"text-align:center; padding-top:5px; border:none;\"><input type=\"hidden\" name=\"openotp_u2f\" value=\"\">'";
				if( $otpChallenge ){		
					$overlay .= "+ '<b>U2F response</b> &nbsp; <blink id=\"u2f_activate\">[Activate Device]</blink></td></tr>'";
				} else { 
					$overlay .= "+ '<img src=\"" . $this->home . "/u2f.png\"><br><br><blink id=\"u2f_activate\">[Activate Device]</blink></td></tr>'";
				}			
			}		
			
			$overlay .= <<<EOT
			+ '</table></form>';
			
			document.body.appendChild(overlay_bg);    
			document.body.appendChild(overlay); 
		}
		
		addOpenOTPDivs();
		
		/* Compute Timeout */	
		var c = $timeout;
		var base = $timeout;
		function count()
		{
			plural = c <= 1 ? "" : "s";
			document.getElementById("timeout").innerHTML = c + " second" + plural;
			var div_width = 360;
			var new_width =  Math.round(c*div_width/base);
			document.getElementById('div_orange').style.width=new_width+'px';
			
			if(c == 0 || c < 0) {
				c = 0;
				clearInterval(timer);
				document.getElementById("timout_cell").innerHTML = " <b style='color:red;'>Login timedout!</b> ";
				document.getElementById("inputs_cell").innerHTML = "<input style='padding:3px 20px;' type='button' value='Retry' class='button mainaction' onclick='window.location.href=\"./\"'>";
			}
			c--;
		}
		count();
		
		
		function getInternetExplorerVersion() {
		
			var rv = -1;
		
			if (navigator.appName == "Microsoft Internet Explorer") {
				var ua = navigator.userAgent;
				var re = new RegExp("MSIE ([0-9]{1,}[\.0-9]{0,})");
				if (re.exec(ua) != null)
					rv = parseFloat(RegExp.$1);
			}
			return rv;
		}
		
		var ver = getInternetExplorerVersion();
		
		if (navigator.appName == "Microsoft Internet Explorer"){
			if (ver <= 10){
				toggleItem = function(){
					
				    var el = document.getElementsByTagName("blink")[0];
				    if (el.style.display === "block") {
				        el.style.display = "none";
				    } else {
				        el.style.display = "block";
				    }
				}
				var t = setInterval(function() {toggleItem; }, 1000);
			}
		}
		
		var timer = setInterval(function() {count();  }, 1000);
EOT;


		if( $u2fChallenge ){ 
			$overlay .= " jQuery(document).ready(function(){ " . "\r\n";
			$overlay .= "if (/chrome|chromium|firefox|opera/.test(navigator.userAgent.toLowerCase())) {
			    var u2f_request = ".$u2fChallenge.";
			    var u2f_regkeys = [];
			    for (var i=0, len=u2f_request.keyHandles.length; i<len; i++) {
			        u2f_regkeys.push({version:u2f_request.version,keyHandle:u2f_request.keyHandles[i]});
			    }
			    u2f.sign(u2f_request.appId, u2f_request.challenge, u2f_regkeys, function(response) {
					document.getElementsByName('openotp_u2f')[0].value = JSON.stringify(response); 
					document.getElementById('login-form-otp').submit();					
			    }, $timeout ); }" . "\r\n";
			$overlay .= " else { 
				var u2f_activate = document.getElementById('u2f_activate'); 
				u2f_activate.innerHTML = '[Not Supported]'; 
				u2f_activate.style.color='red'; 
				}" . "\r\n";
			$overlay .= " }); " . "\r\n";			
		}
		
		
		//if( $u2fChallenge ) $overlay .= " if (typeof u2f !== 'object' || typeof u2f.sign !== 'function'){ var u2f_activate = document.getElementById('u2f_activate'); u2f_activate.innerHTML = '[Not Supported]'; u2f_activate.style.color='red'; }" . "\r\n";
		//if( $u2fChallenge ) $overlay .= " else {  u2f.sign([".$u2fChallenge."], function(response) { document.getElementsByName('openotp_u2f')[0].value = JSON.stringify(response); document.getElementById('login-form-otp').submit(); }, $timeout ); }" . "\r\n";

		return $overlay;
	}
	
	private function soapRequest(){
	
		$options = array('location' => $this->server_url);
		if ($this->proxy_host != NULL && $this->proxy_port != NULL) {
			$options['proxy_host'] = $this->proxy_host;
			$options['proxy_port'] = $this->proxy_port;
			if ($this->proxy_username != NULL && $this->proxy_password != NULL) {
				$options['proxy_login'] = $this->proxy_username;
				$options['proxy_password'] = $this->proxy_password;
			}
		}
		
		$stream_context = stream_context_create(array('ssl' => array('verify_peer' => false)));
		if ($stream_context){
			$options['stream_context'] = $stream_context;
		}
				
		ini_set('soap.wsdl_cache_enabled', '0');
		ini_set('soap.wsdl_cache_ttl', '0'); 
		
		try {
			$soap_client = new SoapClientTimeout(dirname(__FILE__).'/openotp.wsdl', $options);				
		} catch (Exception $e) {
			return false;
		}

		if (!$soap_client) {
			return false;
		}
		$soap_client->setTimeout(30);		
		$soap_client->setVersion(2);		
		$this->soap_client = $soap_client;	
		
		return true;
	}
		
	public function openOTPSimpleLogin($username, $domain, $password, $remote_add, $context){
		echo $user_settings;
		if (!$this->soapRequest()) return false;
		$resp = $this->soap_client->openotpSimpleLogin($username, $domain, $password, $this->client_id, $remote_add, $this->user_settings, NULL, $context);
		
		return $resp;
	}
	
	public function openOTPChallenge($username, $domain, $state, $password, $u2f){
		if (!$this->soapRequest()) return false;
		$resp = $this->soap_client->openotpChallenge($username, $domain, $state, $password, $u2f);
		
		return $resp;
	}
}

if (extension_loaded('soap')) {
class SoapClientTimeout extends SoapClient {
    private $timeout;
    private $version;

    public function setTimeout ($timeout) {
        $this->timeout = $timeout;
    }
    public function setVersion ($version) {
        $this->version = $version;
    }

    public function __doRequest($request, $location, $action, $version, $one_way=false) {
        if (!$this->timeout) {
            // Call via parent because we require no timeout
            $response = parent::__doRequest($request, $location, $action, $version, $one_way);
        } else {
            // Call via Curl and use the timeout
            $curl = curl_init($location);

            curl_setopt($curl, CURLOPT_VERBOSE, false);
            curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($curl, CURLOPT_POST, true);
            curl_setopt($curl, CURLOPT_POSTFIELDS, $request);
            curl_setopt($curl, CURLOPT_HEADER, false);
            curl_setopt($curl, CURLOPT_HTTPHEADER, array("Content-Type: text/xml", "API-Version: ".strval($this->version)));
            curl_setopt($curl, CURLOPT_TIMEOUT, $this->timeout);

            $response = curl_exec($curl);
            if (curl_errno($curl)) throw new Exception(curl_error($curl));
            curl_close($curl);
        }

        if (!$one_way) return ($response);
    }
}
}else{
	return false;
}

?>

openotp_authentication/openotp_authentication.php
<?php
/*
 RCDevs OpenOTP Plugin for RoundCube Webmail v1.2.3
 Copyright (c) 2010-2016 RCDevs, All rights reserved.
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
  
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

class openotp_authentication extends rcube_plugin {
    public $task = 'login';
    private $state = NULL;
    private $message = NULL;
    private $timeout = NULL;
    private $ldappw = NULL;
    private $domain = NULL;
    private $username = NULL;
	private $error = NULL;
	private $openotp_auth = NULL;
    

    public function init() {
		$this->load_config();
	    $this->add_hook('startup', array($this, 'startup'));
		$this->add_hook('authenticate', array($this, 'authenticate'));
		$this->add_hook('template_object_loginform', array($this, 'login_form'));
	    $rcmail = rcmail::get_instance();
		$this->rc = $rcmail;
		
	}
	

    public function startup($args)
    {
		
        $include_path = $this->home . PATH_SEPARATOR;
        $include_path .= ini_get('include_path');
        set_include_path($include_path);

		$this->openotp_auth = new openotp($this, "plugins/openotp_authentication");
		
		// Check if config files are available		
		if (!$this->openotp_auth->checkFile('config.inc.php','No OpenOTP config file found')){
			write_log('errors', 'No OpenOTP config file found');
			$this->error = 'No OpenOTP config file found';
			return $args;
		}
		if (!$this->openotp_auth->checkFile('openotp.wsdl','Could not load OpenOTP WSDL file')){
			write_log('errors', 'Could not load OpenOTP WSDL file');
			$this->error = 'Could not load OpenOTP WSDL file';
			return $args;
		}
		// Check SOAP extension is loaded
		if (!$this->openotp_auth->checkSOAPext()){
			write_log('errors', 'Your PHP installation is missing the SOAP extension');
			$this->error = 'Your PHP installation is missing the SOAP extension';
			return $args;
		}
		// require server_url
		if (!$this->openotp_auth->getServer_url()) {
			write_log('errors', 'OpenOTP server URL is not configured');
			$this->error = 'OpenOTP server URL is not configured';
			return $args;
		}
		
        return $args;
    }


    public function login_form($form) {
		
		$rcmail = rcmail::get_instance();

		if ($this->username != NULL && $this->state != NULL && $this->ldappw != NULL){
			$otp_script = $this->openotp_auth->getOverlay($this->otpChallenge, $this->u2fChallenge, $this->message, $this->username, $this->state, $this->timeout, $this->ldappw, $this->domain);
			//$rcmail->output->add_header('<script type="text/javascript" src="chrome-extension://pfboblefjcgdjicmnffhdgionmgcdmne/u2f-api.js"></script>');
			$rcmail->output->add_header('<script type="text/javascript" src="plugins/openotp_authentication/fidou2f.js"></script>');
			$rcmail->output->add_script($otp_script, 'docready');
		}		

		$inline_js = "
			jQuery('#login-form form').submit(function () {
				jQuery(this).prepend(\"<span style='color:white; text-shadow: 0px 1px 1px black; font-size:0.9em;'>Processing request. Please wait...</span>\");
				return true;
			});";

		$rcmail->output->add_script($inline_js, 'docready');
		
		$table = new html_table(array('cols' => 1));
		if ($this->message != NULL) $table->add('title', html::label('openotmessage', '<font color=red>'.$this->message.'</font>'));
		if ($this->error != NULL) $table->add('title', html::label('openotmessage', '<font color=red><b>'.$this->error.'</b></font>'));
		$form['content'] = $table->show().$form['content'];
		
		return $form;
    }
	
	
    
	public function authenticate($data) {

		if ($this->error != NULL){
			$data['valid'] = false;
			return $data;
		}

		// Get context cookie
		$context_name = $this->openotp_auth->getContext_name();
		$context_size = $this->openotp_auth->getContext_size();
		$context_time = $this->openotp_auth->getContext_time();
		
		if (isset($_COOKIE[$context_name])) $context = $_COOKIE[$context_name];
		else $context = NULL;	
		
		$username = get_input_value('openotp_username', RCUBE_INPUT_POST) != NULL ? get_input_value('openotp_username', RCUBE_INPUT_POST) : $data['user'];
		$password = get_input_value('openotp_password', RCUBE_INPUT_POST) != NULL ? get_input_value('openotp_password', RCUBE_INPUT_POST) : $data['pass'];
		$u2f = $_POST['openotp_u2f'] != NULL ? $_POST['openotp_u2f'] : "";
		$state = get_input_value('openotp_state', RCUBE_INPUT_POST);
		$ldappw = get_input_value('openotp_ldappw', RCUBE_INPUT_POST);

		
		if (empty($username)) {
			$data['valid'] = false;
			return $data;
		}
		
		$t_domain = $this->openotp_auth->getDomain($username);
		if (is_array($t_domain)){
			$username = $t_domain['username'];
			$this->domain = $t_domain['domain'];
		}elseif($_POST['openotp_domain'] != NULL) $this->domain = $_POST['openotp_domain'];
		else $this->domain = $t_domain;

		if (!$this->openotp_auth->enableOpenotp_auth()) {
			write_log('errors', 'Plugin OpenOTP authentication configured and disabled');
			return $data;
		}
		
		if ($state != NULL) {
			if (!$ldappw) {
				write_log('errors', 'No LDAP password provided for user '.$data['user']);
				$this->error = 'No LDAP password provided for user '.$data['user'];
				$data['valid'] = false;                                                                                           
				return $data;
			}
			// OpenOTP Challenge
			$resp = $this->openotp_auth->openOTPChallenge($username, $this->domain, $state, $password, $u2f);
		} else {
			// OpenOTP Login
			$resp = $this->openotp_auth->openOTPSimpleLogin($username, $this->domain, utf8_encode($password), $_SERVER['REMOTE_ADDR'], $context);
			if(!$resp){
				write_log('errors', 'Could not load OpenOTP WSDL file');
				$data['valid'] = false;
				return $data;				
			}
		}
		
		if (!$resp || !isset($resp['code'])) {
			write_log('errors', 'Invalid OpenOTP response for user '.$data['user']);
			$this->error = 'Internal system error, please contact administrator';
			$data['valid'] = false;
			return $data;
		}

		switch ($resp['code']) {
			case 0:
				$this->message = $resp['message'];
				$data['abort'] = true;
				break;
			case 1:
				$data['user'] = $username;
				$data['pass'] = isset($ldappw) ? $ldappw : $password;
				
				// set context cookie
				if (extension_loaded('openssl')) {			
					if (strlen($context) != $context_size) $context = bin2hex(openssl_random_pseudo_bytes($context_size/2));
					setcookie($context_name, $context, time()+$context_time, '/', NULL, true, true);
				}				
				
				$data['valid'] = true;
				break;
			case 2:
				$this->message = $resp['message'];
				$this->state = $resp['session'];
				$this->timeout = $resp['timeout'];
				$this->ldappw = $data['pass'];
				$this->otpChallenge = $resp['otpChallenge'];
				$this->u2fChallenge = $resp['u2fChallenge'];
				$this->username = $username;
				$data['abort'] = true;
				break;
			default:
				write_log('errors', 'Invalid OpenOTP response code '.$resp['code'].' for user '.$data['user']);
				$this->error = 'Internal system error, please contact administrator';
				$data['valid'] = false;
				break;
		}

		return $data;
	}
}

?>